Work on Node class 3 (#119)
CodeYourFuture/syllabus: docs/guides/marking-guide.md

---
id: marking-guide
title: Marking Guide
sidebar_label: Marking Guide
---

This Rubric should be used to guide your assessment of Students Homework at CodeYourFuture

## Overview

### Completeness (/10)

> - The code achieves all of the tasks in the coursework

Reviewer: Again with the homework/coursework distinction - let's aim for one term if we can?

> - The code achieves all of the tasks in the coursework

Reviewer: May be worth calling out something about it doing so in the intended way. Sometimes we see learners e.g. changing test cases, or over-fitting their solutions to the test cases in the homework - they should aim to identify and solve the general problem set by the task.

- The code handles error responses as well as success responses [[read more]](#error-responses)

### Code Style (/5)

- The code follows a consistent style
- The code follows CodeYourFuture's coding standards [[read more]](#coding-standards)
- The code is well indented [[read more]](#well-indented-code)
- The code uses good variable names [[read more]](#good-variable-names)
- The code uses good function names [[read more]](#good-function-names)
- The code uses more optimal, less error prone code to solve problems [[read more]](#more-optimal-code)
- The code should not use hard coded values [[read more]](#hard-coded-values)

### Clarity (/5)

- The code is divided into functions of a good size. [[read more]](#small-functions)
- Code is not repeated in several locations of the coursework. [[read more]](#non-duplicated-code)
- The coursework does not have commented out code. [[read more]](#commented-out-code)
- There is no dead code in the coursework [[read more]](#dead-code)
- There is no extra or debug logging [[read more]](#debug-logging)

## Explanations

### Edge Cases

Edge cases are when code does something you're not expecting and you didn't account for in your code.

> For example, using a number in a `Name` field of a form might cause problems since peoples names can't be numbers.

Reviewer: ```suggestion
Some edge cases are valid data, and some are not.

Some examples of valid edge case are that some names contain punctuation, or some people may only have a single name, which your code needs to be able to handle. An example of an invalid edge case is if someone doesn't enter a name at all, which your code should reject - if your code doesn't notice the problem where the user provides their name, other code you wrote may make assumptions which are incorrect.

See [Falsehoods Programmers Believe About Names](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) for some interesting examples.
```

Good code should handle all of these cases so that they can't happen - and if they do - they don't make your program crash or stop work.

Validations are used wherever necessary. The code never breaks under any circumstances. Especially under invalid inputs that come from the user end. Regardless of it being a negative, over-sized, invalid format, etc., every input passed should be processed, sanitized before taking it further. Every object is checked for its actual data existence before accessing its properties.

#### More Reading

- [Don't forget edge cases](https://www.geeksforgeeks.org/dont-forget-edge-cases/)

### Small Functions

> Writing small functions makes your code easier to understand and easier to work with

Reviewer: ```suggestion
Writing small functions makes your code easier to understand and easier to work with. Ideally you should be able to describe what a function does in one sentence, without the word "and".

Good examples: "This function returns the user's favourite author", "This other function Tweets the title of an author's most popular book".

Bad example: "This function finds the user's favourite author and sends a Tweet with the name of that author's most popular book".
```

#### More Reading

- [How big should a function be?](https://thecodersblog.com/how-big-a-function-should-be/)

### Debug Logging

The code should not contain debug logging that was used during development. For example, any logs like

```js
console.log("Does this work");
console.log("1");
console.log(array[1].object.name);
```

should be removed.

### Error Responses

> Not just the error messages, every response that is returned by the server must be properly handled. It should have necessary headers, response messages, error codes and any other necessary details attached with it in required format. All possible scenarios are tested to avoid deadlocks, timeouts, etc.

Reviewer: ```suggestion
Not just the error messages, every response that is returned by the server must be properly handled. It should have necessary headers, response messages, error codes and any other necessary details attached with it in required format. All possible scenarios are tested to avoid timeouts, unresponsive UI, etc.
```

I feel like deadlocks are a more advanced topic than we cover :)

#### More Reading

- [Handling Error HTTP Responses in Javascript fetch](https://usefulangle.com/post/314/javascript-fetch-error-handling)

### Non-duplicated Code

Duplicated code should be moved into functions so they can be reference from multiple places.

Having lots of duplicated code in your coursework means that it is harder to change or update your code.

#### More Reading

- [Duplication Code](https://dev.to/thinkster/code-smell-duplicated-code-fc1)

> ### Commented Out Code

Reviewer: Maybe worth plugging git history here?

Commenting out code is a great way to debug your code but when you use it to keep code for later it can become a problem.

Instead of commenting out code you want to keep but not use, consider moving it to separate file. This will make it easier to keep track of the code you're currently working on.

Having lots of commented out code make your code harder to read and make it easier for you to get overwhelmed and lost in your code.

#### More Reading

- [Please don't comment out code](https://kentcdodds.com/blog/please-dont-commit-commented-out-code)
- [Putting comments in code: the good, the bad, and the ugly.](https://www.freecodecamp.org/news/code-comments-the-good-the-bad-and-the-ugly-be9cc65fbf83/)

### Dead Code

Dead Code is code that is in your coursework but can't be reached or run by your program.

> For example, a function that is never called (or invoked) is considered dead code.

Reviewer: ```suggestion
For example, a function that is never called is considered dead code.
```

#### More Reading

- [CodeYourFuture Coding Standards](./coding-standards)

### Well Indented Code

Indentation is a fundamental aspect of code styling, and plays a large role in influencing readability. First of all, indented code is easier to read through than unindented code.

With unindented code, the overall structure of the code might be somewhat difficult to see. However, with indented code, the overall structure of the code jumps out at you. Tabs tell you that a line of code is inside a function, loop, if statement, or else statement. Knowing which parts of the code is inside what will become especially important when your code becomes more complex.

#### More Reading

- [Importance of Code Indentation](http://mrbool.com/importance-of-code-indentation/29079)

### Good Function Names

Good function names are important so that code is easy to read and understand what it is doing.

A bad function name might be

```js
function doWork() {}
```

A good function name might be

```js
function removeCaptialLetters(){)
```

> Sometimes if you can't come up with a good function name it's an indication that the function doesn't have a nice, crisp focus and needs to be refactored. If it's a class method, perhaps the class needs refactoring too.

Reviewer: ```suggestion
Sometimes if you can't come up with a good function name it's an indication that the function doesn't have a nice, crisp focus and needs to be refactored.
```

We don't teach classes :)

Also, I'm not sure we ever teach/explain the term refactoring?

But it's well worth the trouble finding the best possible names, since it makes your code so much more understandable and usable.

#### More Reading

- [Coding like Shakespeare: Practical Function Naming Conventions](https://dmitripavlutin.com/coding-like-shakespeare-practical-function-naming-conventions/)
- [Trouble coming up with good names for functions](https://stackoverflow.com/questions/979118/trouble-coming-up-with-good-names-for-functions)

### Good Variable Names

Each variable is named so it is clear which variable is being used at any time. It is important to use meaningful names for variables:

For example,

```js
let pocketMoney = 20;
```

> means that the variable ‘pocketMoney’ is being used to store how much pocket money you have. Right now you have £20.

Reviewer: ```suggestion
means that the variable `pocketMoney` is being used to store how much pocket money you have. Right now you have £20.
```

> The name given to each variable is up to the programmer, but ideally a variable name should have meaning, ie it should reflect the value that it is holding.

Reviewer: ```suggestion
The name given to each variable is up to the programmer, but ideally a variable name should have meaning you can understand without needing to read any other code, i.e. it should reflect the value that it is holding.
```

#### More Reading

- [Good Variable Names](https://spin.atomicobject.com/2017/11/01/good-variable-names/)
- [Programming Basics - Naming Variables](https://www.bbc.co.uk/bitesize/guides/zwmbgk7/revision/3)

> ### More Optimal Code

Reviewer: I'm not sure about this section for a couple of reasons

1. Optimal in my mind means efficient, not readable
2. Our learners should favour readability over efficiency most of the time
3. We don't really teach anything about efficiency.

I feel like I'd rather frame this as something like:

> ### More readable code
> 
> The most important thing when writing code is to make it easy to read and understand, and as simple as possible.
> 
> [contrasting examples of `filter` and `for` go here - ideally two examples, one showing a case where `filter` is simpler, one showing a case where a `for` loop is simpler]

(And update the bullet which links here to talk about readability not optimality)

If we want to talk efficiency, we can do that too, but I suspect we don't.

Reviewer 2: That's exactly what I was going for - thanks!

Nearly always, there are better and worse ways of writing code. You should always try to write code in a way that is simple to understand and read.

An example of write more optimal code might be using a `filter` instead of a complicated `for` loop. This is better, because it is immediately obvious what you're trying to do - filter an array - and you're less likely to make a mistake using it.

### Hard Coded Values

Hard coded values with no explanation make code hard to understand and read in the long term. For example

```js
getData("https://www.api.com/retrieve");
calculateScreenSize(1800);
```

Will be hard to understand when you codebase grows.

> Instead us constants to store your hard coded values. For example

Reviewer: ```suggestion
Instead us constants to store your hard coded values - this means the value has a name, which can be [usefully descriptive](#good-variable-names). For example
```

> Test 3: let array = ["Up", "Down"]

To make this test pass you could write

```js
console.log(array[0]);
console.log(array[1]);
console.log(array[2]);
console.log(array[3]);
```

However, that would not work for _every_ case that exists - only the ones that we're writing. Your aim isn't to just pass the tests, it's to write good code that solves the problem.

A better solution would be

```js
array.forEach((item) => {
  console.log(item);
});
```

### Edge Cases

Edge cases are when code does something you're not expecting and you didn't account for in your code.

Some edge cases are valid data, and some are not.

Some examples of valid edge case are that some names contain punctuation, or some people may only have a single name, which your code needs to be able to handle. An example of an invalid edge case is if someone doesn't enter a name at all, which your code should reject - if your code doesn't notice the problem where the user provides their name, other code you wrote may make assumptions which are incorrect.

See [Falsehoods Programmers Believe About Names](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) for some interesting examples.

> Good code should handle all of these cases so that they can't happen - and if they do - they don't make your program crash or stop work.

Reviewer: ```suggestion
Good code should handle all of these cases so that they can't happen - and if they do - they don't make your program crash or stop working.
```

For example,

```js
const trees = ["Oak", "Birch", "Beech", "Elm", "Hazel"];

let treesBeginningWithB = [];
for (var i = 0; i < trees.length; i++) {
  if (trees[i] === "B") {
    treesBeginningWithB.push(trees[i]);
  }
}

console.log(treesBeginningWithB); // ["Birch", "Beech"]
```

Could be written like this be easier to understand and read.

```js
const trees = ["Oak", "Birch", "Beech", "Elm", "Hazel"];

let treesBeginningWithB = trees.filter((item) => item.chatAt(0) === "B");

console.log(treesBeginningWithB); // ["Birch", "Beech"]
```

As you can see, we've reduced six lines of code into one and it is immediately obvious that it is trying to filter the trees beginning with B.

Writing readable code is a big topic and one that you'll get better at as you go on. We encourage you to read more about it.

#### More Reading

- [Readable Code](https://blog.pragmaticengineer.com/readable-code/)

### Hard Coded Values

Hard coded values with no explanation make code hard to understand and read in the long term. For example

```js
getData("https://www.api.com/retrieve");
calculateScreenSize(1800);
```

Will be hard to understand when you codebase grows.

> Instead us constants to store your hard coded values - this means the value has a name, which can be [usefully descriptive](#good-variable-names). For example

Reviewer: ```suggestion
Instead use constants to store your hard coded values - this means the value has a name, which can be [usefully descriptive](#good-variable-names). For example
```

```js

> let ENDPOINT = "https://www.api.com/retrieve";

Reviewer: ```suggestion
let WEATHER_ENDPOINT = "https://www.api.com/retrieve";
```

Let's be even less generic :)

> let WIDTH = 1800;

Reviewer: ```suggestion
let MAX_SCREEN_WIDTH = 1800;
```
